src/work/Work.h

```diff
@@ -1,140 +1,73 @@
-    // WORK_SUCCESS, but it also returns the next sate desired: if you want
-    // to restart or keep going, return WORK_PENDING or WORK_RUNNING
-    // (respectively) from onSuccess and you'll be rescheduled to run more.
-    // If you want to force failure (and reset / retry) you can return
-    // WORK_FAILURE_RETRY or WORK_FAILURE_RAISE. After a retry count is
-    // passed, WORK_FAILURE_RETRY means WORK_FAILURE_RAISE anyways.
-    // WORK_FAILURE_FATAL is equivalent to WORK_FAILURE_RAISE passed up in
-    // the work chain - when WORK_FAILURE_FATAL is raised in one work item,
-    // all work items that leaded to this one will also fail without retrying.
-    virtual State onSuccess();
-
-    static std::string stateName(State st);
-    State getState() const;
-    bool isDone() const;
-    void advance();
-    void reset();
-
-  protected:
-    std::weak_ptr<WorkParent> mParent;
-    std::string mUniqueName;
-    size_t mMaxRetries{RETRY_A_FEW};
-    size_t mRetries{0};
-    State mState{WORK_PENDING};
-    bool mScheduled{false};
-
-    std::unique_ptr<VirtualTimer> mRetryTimer;
-
-    std::function<void(asio::error_code const& ec)> callComplete();
-    void run();
-    void complete(CompleteResult result);
-    void scheduleComplete(CompleteResult result = WORK_COMPLETE_OK);
-    void scheduleRetry();
-    void scheduleRun();
-    void
-    scheduleSuccess()
+    template <typename T, typename... Args>
+    std::shared_ptr<T>
+    addWork(Args&&... args)
     {
-        scheduleComplete();
+        // `wakeUp` is sufficient as a callback for any child work of
+        // WorkScheduler (see custom callback for WorkScheduler)
+        std::weak_ptr<Work> weak(
+            std::static_pointer_cast<Work>(shared_from_this()));
+        auto callback = [weak]() {
+            auto self = weak.lock();
+            if (self)
+            {
+                self->wakeUp();
+            }
+        };
+
+        auto child =
+            std::make_shared<T>(mApp, callback, std::forward<Args>(args)...);
+        addChild(child);
+        wakeUp();
+        return child;
     }
-    void
-    scheduleFailure()
-    {
-        scheduleComplete(WORK_COMPLETE_FAILURE);
-    }
-    void
-    scheduleFatalFailure()
-    {
-        scheduleComplete(WORK_COMPLETE_FATAL);
-    }
-
-    void setState(State s);
 
-    void notifyParent();
-    virtual void notify(std::string const& childChanged) override;
+    // Children status helper methods
+    bool allChildrenSuccessful() const;
+    bool allChildrenDone() const;
+    bool anyChildRaiseFailure() const;
+    bool anyChildFatalFailure() const;
+    bool anyChildRunning() const;
+    bool hasChildren() const;
 
-  private:
-    VirtualClock::duration getRetryDelay() const;
+  protected:
+    Work(Application& app, std::function<void()> callback, std::string name,
+         size_t retries = BasicWork::RETRY_A_FEW);
+    void clearChildren();
```

Comment:
Investigated this more and talked to @MonsieurNicolas, and it appears that the original reason behind calling `clearChildren` in `Work*` destructors in an interesting case of `shared_ptr` vs resource de-allocation (this commit might be helpful: https://github.com/stellar/stellar-core/pull/1355). Essentially, it seems like a child Work class was able to get a `shared_ptr` from `weak_ptr` to its parent during that parent's destruction and call its methods, causing a crash. Looks like the fix was to `clearChildren` first to prevent calls to parent. 
Now, with the new design, we'd like to avoid this altogether, and I think it is possible with proper shutdown of `WorkScheduler`. The idea is to provide a guarantee that no work would not be destroyed until it reached complete state and reset (which implies cleaning up references to its children). This is basically an abort functionality. While the goal is to have it implemented once the implementation for the new work design is in place, for this first iteration, we decided to try and add WORK_DESTRUCTING state, and perform a no-op in `wakeUp` (as any child work would have a weak_ptr to parent in its callback and could call `wakeUp` on its half-destroyed parent).